arch/xtensa/src/esp32/esp32_wifi_adapter.c

```diff
@@ -3447,7 +3469,17 @@ static void *esp_malloc_internal(size_t size)
 
 static void *esp_realloc_internal(void *ptr, size_t size)
 {
+#ifdef CONFIG_XTENSA_IMEM_USE_SEPARATE_HEAP
+  return xtensa_imm_realloc(ptr, size);
+#else
+  if (size == 0 || esp32_ptr_extram(ptr))
+    {
+      esp_free(ptr);
+      return NULL;
```

Comment:
* size == 0 case is better to be left to kmm_realloc
* standard realloc is supposed to preserve the old memory in case of a failure. does this wifi_osi_funcs_t variant have a different semantics?

Reply:
The size == 0 case means that the size of the new malloc is 0,
In the esp_realloc_internal processing of wifi_osi_funcs_t, you can refer to:
https://github.com/espressif/esp-idf/blob/master/components/heap/heap_caps.c#L318-L321

Reply:
the esp-idf implementation does not seem to free the given `ptr` in case of error.
eg. https://github.com/espressif/esp-idf/blob/master/components/heap/heap_caps.c#L323-L327
it matches my expectation for realloc().
i guess this implementation should do the same.
unfortunately it seems ~impossible to implement it for !CONFIG_XTENSA_IMEM_USE_SEPARATE_HEAP case though.

Reply:
Yes, as the heap management of NuttX and esp-idf is too different, this part of the modification is too big, now temporarily free the given `ptr` in case of error.

Reply:
but doesn't it cause double-free sooner or later, as the library would think the `ptr` still needs to be freed?

Reply:
maybe you can do something like the following instead of kmm_realloc. how do you think?
```
new_ptr = kmm_malloc(new_size);
if (esp32_ptr_extram(new_ptr)) {
    kmm_free(new_ptr);
    return NULL;
}
old_size = malloc_usable_size(old_ptr);
memcpy(new_ptr, old_ptr, min(old_size, new_size));
kmm_free(old_ptr);
return new_ptr
```

Reply:
@yamt The library will check `ptr` and then free it, there is a risk of double-free and I also think the above code instead of kmm_realloc is more reasonable, I have modified it, thank you.